[Functional Test Flakiness] Improve Reliability of AWS Resource Validation

### Area for Improvement

AWS functional tests

### Observed behavior

We have a functional test that creates an AWS S3 bucket, validates its existence by performing a `get` request, returns an error if the get call results in an error (including 404), and finally performs a cleanup step to delete the resources, even if the previous step returned an error. In a recent run, this test failed during the validation of the S3 bucket creation because the bucket could not be found. Logs -

```
cli.go:341: [rad]     radiusfunctionaltestbucket-aca3abad-f25b-49d3-be9f-bd1f7f543171 AWS.S3/Bucket       
    cli.go:341: [rad] 
    deployexecutor.go:84: finished deploying deploy testdata/aws-s3-bucket.bicep from file testdata/aws-s3-bucket.bicep
    rptest.go:3[93](https://github.com/radius-project/radius/actions/runs/11243305588/job/31259088130#step:26:94): finished running step 0 of 1: deploy testdata/aws-s3-bucket.bicep
    rptest.go:396: skipping validation of resources...
    rptest.go:411: validating output resources for deploy testdata/aws-s3-bucket.bicep
    aws.go:72: 
        	Error Trace:	/home/runner/work/radius/radius/test/validation/aws.go:72
        	            				/home/runner/work/radius/radius/test/rp/rptest.go:413
make: *** [build/test.mk:89: test-functional-corerp-cloud] Error 1
        	Error:      	Received unexpected error:
        	            	operation error CloudControl: GetResource, https response error StatusCode: 400, RequestID: b546bee7-401d-4efe-9bc8-7a9ecd041e27, ResourceNotFoundException: AWS::S3::Bucket Handler returned status FAILED: Bucket not found (HandlerErrorCode: NotFound, RequestToken: a64[102](https://github.com/radius-project/radius/actions/runs/11243305588/job/31259088130#step:26:103)cb-8ab2-407d-8330-e85842af474e)
        	Test:       	Test_AWS_S3Bucket/deploy_testdata/aws-s3-bucket.bicep
    --- FAIL: Test_AWS_S3Bucket/deploy_testdata/aws-s3-bucket.bicep (40.26s)
```

It errored out again during cleanup stage unable to validate deletion of the bucket -

```
rptest.go:458: validating deletion of AWS resource for radiusfunctionaltestbucket-aca3abad-f25b-49d3-be9f-bd1f7f543171 (attempt 5/5)
    rptest.go:475: 
        	Error Trace:	/home/runner/work/radius/radius/test/rp/rptest.go:475
        	Error:      	Should be true
        	Test:       	Test_AWS_S3Bucket
        	Messages:   	AWS resource radiusfunctionaltestbucket-aca3abad-f25b-49d3-be9f-bd1f7f543171 was present, should be not found

DONE 20 tests, 2 skipped, 2 failures in 151.163s
```

Link to functional test run: https://github.com/radius-project/radius/actions/runs/11243305588/job/31259088130

### Desired behavior

The test should always pass.

### Proposed Fix

1. From the logs above, it's clear that the bucket was eventually created. So adding retries with backoff on the validation path [here](https://github.com/radius-project/radius/blob/main/test/validation/aws.go#L68) should help mitigate intermittent failures.

2. The second failure happened because the bucket wasn't deleted within the time allocated for the test validation. We should consider increasing the max retry limit [here](https://github.com/radius-project/radius/blob/b2eb80b021abb238ae9b93496601ea665f634ccf/test/rp/rptest.go#L457). 

3. The backoff logic could be improved as well - currently there is a fixed 10 second forced wait between retries. We could start with a smaller wait time  and exponentially increase it for cases where operations are delayed due to external issues. Since the buckets created for this test don't contain any objects, the creation and deletion should be fairly quick in most cases.

### rad Version

N/A

### Operating system

N/A

### Additional context

_No response_

### Would you like to support us?

- [ ] Yes, I would like to support you

[AB#13454](https://dev.azure.com/azure-octo/Incubations/_workitems/edit/13454)

--- a/test/rp/rptest.go
+++ b/test/rp/rptest.go
@@ -454,6 +454,8 @@ func (ct RPTest) Test(t *testing.T) {
 
 					// Ensure that the resource is deleted with retries
 					notFound := false
+					baseWaitTime := 15 * time.Second
+
 					for attempt := 1; attempt <= AWSDeletionRetryLimit; attempt++ {
 						t.Logf("validating deletion of AWS resource for %s (attempt %d/%d)", ct.Description, attempt, AWSDeletionRetryLimit)
 
@@ -467,8 +469,10 @@ func (ct RPTest) Test(t *testing.T) {
 							t.Logf("checking existence of resource %s failed with err: %s", resource.Name, err)
 							break
 						} else {
-							// Wait for 10 seconds
-							time.Sleep(10 * time.Second)
+							// Wait with exponential backoff
+							waitTime := baseWaitTime * time.Duration(attempt)
+							t.Logf("waiting for %s before next attempt", waitTime)
+							time.Sleep(waitTime)
 						}
 					}
 

--- a/test/validation/aws.go
+++ b/test/validation/aws.go
@@ -61,14 +61,28 @@ type AWSResourceSet struct {
 
 // ValidateAWSResources checks that the expected AWS resources exist and have the expected properties.
 func ValidateAWSResources(ctx context.Context, t *testing.T, expected *AWSResourceSet, client awsclient.AWSCloudControlClient) {
+	baseWaitTime := 15 * time.Second
+	retryLimit := 5
+
 	for _, resource := range expected.Resources {
 		resourceType, err := GetResourceTypeName(ctx, &resource)
 		require.NoError(t, err)
 
-		resourceResponse, err := client.GetResource(ctx, &cloudcontrol.GetResourceInput{
-			Identifier: to.Ptr(resource.Identifier),
-			TypeName:   &resourceType,
-		})
+		var resourceResponse *cloudcontrol.GetResourceOutput
+		for attempt := 1; attempt <= retryLimit; attempt++ {
+			resourceResponse, err = client.GetResource(ctx, &cloudcontrol.GetResourceInput{
+				Identifier: to.Ptr(resource.Identifier),
+				TypeName:   &resourceType,
+			})
+			if err == nil {
+				break
+			}
+
+			t.Logf("attempt %d/%d: failed to get resource %s with error: %s", attempt, retryLimit, resource.Identifier, err)
+			waitTime := baseWaitTime * time.Duration(attempt)
+			t.Logf("waiting for %s before next attempt", waitTime)
+			time.Sleep(waitTime)
+		}
 		require.NoError(t, err)
 
 		if resource.Properties != nil {

